File: packages/elixir/lib/kreuzberg/citation_metadata.ex
defmodule Kreuzberg.CitationMetadata do
  @moduledoc """
  Citation file metadata (RIS, PubMed, EndNote).

  Matches the Rust `CitationMetadata` struct.

  ## Fields

    * `:citation_count` - Number of citations
    * `:format` - Citation format (e.g., "RIS", "PubMed")
    * `:authors` - List of author names
    * `:year_range` - Year range of citations
    * `:dois` - List of DOIs
    * `:keywords` - List of keywords
  """

  @type t :: %__MODULE__{
          citation_count: non_neg_integer(),
          format: String.t() | nil,
          authors: list(String.t()),
          year_range: Kreuzberg.YearRange.t() | nil,
          dois: list(String.t()),
          keywords: list(String.t())
        }

  defstruct citation_count: 0,
            format: nil,
            authors: [],
            year_range: nil,
            dois: [],
            keywords: []

  @spec from_map(map()) :: t()
  def from_map(data) when is_map(data) do
    %__MODULE__{
      citation_count: data["citation_count"] || 0,
      format: data["format"],
      authors: data["authors"] || [],
      year_range: normalize_year_range(data["year_range"]),
      dois: data["dois"] || [],
      keywords: data["keywords"] || []
    }
  end

  @spec to_map(t()) :: map()
  def to_map(%__MODULE__{} = meta) do
    %{
      "citation_count" => meta.citation_count,
      "format" => meta.format,
      "authors" => meta.authors,
      "year_range" => serialize_year_range(meta.year_range),
      "dois" => meta.dois,
      "keywords" => meta.keywords
    }
  end

  defp normalize_year_range(nil), do: nil
  defp normalize_year_range(%Kreuzberg.YearRange{} = yr), do: yr
  defp normalize_year_range(map) when is_map(map), do: Kreuzberg.YearRange.from_map(map)
  defp normalize_year_range(_), do: nil

  defp serialize_year_range(nil), do: nil
  defp serialize_year_range(%Kreuzberg.YearRange{} = yr), do: Kreuzberg.YearRange.to_map(yr)
  defp serialize_year_range(other), do: other
end

File: kreuzberg/packages/php/src/Types/Metadata/CitationMetadata.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Types\Metadata;

/**
 * Citation file metadata (RIS, PubMed, EndNote).
 *
 * Contains information about citation records including format,
 * authors, year ranges, DOIs, and keywords.
 */
readonly class CitationMetadata
{
    /**
     * @param string[] $authors
     * @param string[] $dois
     * @param string[] $keywords
     */
    public function __construct(
        public int $citationCount = 0,
        public ?string $format = null,
        public array $authors = [],
        public ?YearRange $yearRange = null,
        public array $dois = [],
        public array $keywords = [],
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        $rawCitationCount = $data['citation_count'] ?? 0;
        $citationCount = is_int($rawCitationCount) ? $rawCitationCount : (is_numeric($rawCitationCount) ? (int) $rawCitationCount : 0);

        /** @var string|null $format */
        $format = $data['format'] ?? null;

        /** @var string[] $authors */
        $authors = $data['authors'] ?? [];
        if (!is_array($authors)) {
            $authors = [];
        }

        $yearRange = null;
        $yearRangeRaw = $data['year_range'] ?? null;
        if (is_array($yearRangeRaw)) {
            /** @var array<string, mixed> $yearRangeData */
            $yearRangeData = $yearRangeRaw;
            $yearRange = YearRange::fromArray($yearRangeData);
        }

        /** @var string[] $dois */
        $dois = $data['dois'] ?? [];
        if (!is_array($dois)) {
            $dois = [];
        }

        /** @var string[] $keywords */
        $keywords = $data['keywords'] ?? [];
        if (!is_array($keywords)) {
            $keywords = [];
        }

        return new self(
            citationCount: $citationCount,
            format: $format,
            authors: $authors,
            yearRange: $yearRange,
            dois: $dois,
            keywords: $keywords,
        );
    }

    /**
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        $result = [
            'citation_count' => $this->citationCount,
        ];

        if ($this->format !== null) {
            $result['format'] = $this->format;
        }

        if ($this->authors !== []) {
            $result['authors'] = $this->authors;
        }

        if ($this->yearRange !== null) {
            $result['year_range'] = $this->yearRange->toArray();
        }

        if ($this->dois !== []) {
            $result['dois'] = $this->dois;
        }

        if ($this->keywords !== []) {
            $result['keywords'] = $this->keywords;
        }

        return $result;
    }
}
